packages/protobuf-test/src/descriptor-set.test.ts

```diff
@@ -209,6 +215,132 @@ describe("DescriptorSet", () => {
+        }),
+      ).toThrow(
+        /^Edition EDITION_99999_TEST_ONLY is later than the maximum supported edition EDITION_2023$/,
+      );
+    });
+  });
+  describe("edition custom features", () => {
+    const upstream = new UpstreamProtobuf();
+    let testProto: Record<"google/protobuf/unittest_features.proto", string>;
+    let featureSetDefaults: FeatureSetDefaults;
+    beforeAll(async () => {
+      const { dir } = await upstream.getTestProtoInclude();
+      const content = readFileSync(
+        join(dir, "google/protobuf/unittest_features.proto"),
+        "utf-8",
+      );
+      testProto = {
+        "google/protobuf/unittest_features.proto": content,
+      };
+      const bin = await upstream.getFeatureSetDefaults(
+        "PROTO2",
+        "2023",
+        testProto,
+      );
+      featureSetDefaults = FeatureSetDefaults.fromBinary(bin);
+    });
+    test("default values can be read from unknown fields of getFeatures()", async () => {
+      const bin = await upstream.compileToDescriptorSet(
+        {
+          ...testProto,
+          "input.proto": `
+            edition = "2023";
+            package protobuf_unittest;
+            `,
+        },
+        { includeImports: true },
+      );
+      const set = createDescriptorSet(bin, { featureSetDefaults });
+      const file = set.files.find((f) => f.name === "input");
+      expect(file).toBeDefined();
+      if (file !== undefined) {
+        const tf = getTestFeatures(file.getFeatures());
+        expect(tf.intFileFeature).toBe(1);
+      }
+    });
+    test("overrides can be read from unknown fields of getFeatures()", async () => {
+      const bin = await upstream.compileToDescriptorSet(
+        {
+          ...testProto,
+          "input.proto": `
+            edition = "2023";
+            package protobuf_unittest;
+            import "google/protobuf/unittest_features.proto";
+            option features.(pb.test).int_file_feature = 8;
+            `,
+        },
+        { includeImports: true },
+      );
+      const set = createDescriptorSet(bin, { featureSetDefaults });
+      const file = set.files.find((f) => f.name === "input");
+      expect(file).toBeDefined();
+      if (file !== undefined) {
+        const tf = getTestFeatures(file.getFeatures());
+        expect(tf.intFileFeature).toBe(8);
+      }
+    });
+    // test helper to read a message field extension to google.protobuf.FeatureSet
+    function getTestFeatures(
+      features: FeatureSet,
+    ): TestFeatures & Required<TestFeatures> {
+      const uf = TestFeatures.runtime.bin
+        .listUnknownFields(features)
+        .filter((f) => f.no === 9999);
+      assert(uf.length >= 0);
+      const tf = new TestFeatures();
+      for (const { data } of uf) {
+        const bytes = new BinaryReader(data).bytes();
+        tf.fromBinary(bytes, { readUnknownFields: false });
+      }
+      return tf as TestFeatures & Required<TestFeatures>;
+    }
+  });
```

Comment:
Just adding test coverage for language-specific extensions to FeatureSet. They work as expected. getTestFeatures can be replaced when we support extensions (https://github.com/bufbuild/protobuf-es/issues/86).